### PR TITLE
[Bug][Qwen3.5] Honor has_initial_state in fused GDN kernel

### PR DESCRIPTION
**Pull request by qizzzh**
PR #2408 plumbed `has_initial_state` through the chunked / ref GDN paths to fix stale-slot leakage on freshly-allocated mamba slots, but explicitly skipped the fused Pallas kernel ("does not yet honor has_initial_state"). With compact-mamba shrinking the mamba pool to `max_num_reqs+1`, slot reuse is the common case and the FUSED path silently consumes a freed slot's prior state on every brand-new prefill — the same kind of silent corruption that hit the chunked path before PR #2408.

Closes that gap on the FUSED path:

- `fused_gdn_recurrent_kernel.py`: add a `has_initial_state` SMEM input; after the DMA load, `@pl.when(is_new_seq & ~has_init)` overwrites `h_bufs[buf_idx]` with zeros so the recurrent update starts from zero instead of stale tenant data. The decode kernel is unchanged — decodes always have a valid prior state.
- `fused_gdn_kernel_wrapper.py`: forward `has_initial_state` through `fused_gdn`, `_dispatch_with_distribution`, and the `ragged_gated_delta_rule` adapter. Default `None` materializes a static all-True ones array, preserving the pre-fix behaviour for any callers that don't manage slot reuse.
- `gdn_attention.py`: flip the FUSED branch to pass `has_initial_state`, matching the chunked / ref branches.

Perf (apples-to-apples microbench on v7x-8, median of 200 reps, H_qk=2 H_v=8 K=V=128):

  pure_decode_128       2.437 -> 2.447 ms  (+0.4%)
  mixed_64dec_4x256pre  3.668 -> 3.664 ms  (-0.1%)
  pure_prefill_4x1024   5.879 -> 5.877 ms  (-0.0%)
  decode_256            4.237 -> 4.197 ms  (-0.9%)

All within ±1% noise — the SMEM input + conditional adds no measurable cost, and `_zero_h0` only fires on `is_new_seq` for slots flagged False.

Tests (5 new in tests/kernels/fused_gdn_kernel_test.py + 2 FUSED parametrize entries on the existing dispatcher-level `test_has_initial_state_*` tests in tests/layers/common/test_gdn_attention.py; all pass on TPU v7x-8):

- `test_has_initial_state_zeros_stale_slot` — fused-with-stale h0 vs fused-with-fresh h0, the strongest direct test that the kernel zeros stale state.
- `test_has_initial_state_pattern_matches_ref_all_true` — True path preserves the loaded h0.
- `test_has_initial_state_pattern_matches_ref_alternating` — per-slot SMEM lookup (catches off-by-one or uniform-apply bugs).
- `test_has_initial_state_long_sequence_multi_block` — kernel zeros only on `is_new_seq`; subsequent blocks of the same prefill must not re-zero (or the running state gets wiped mid-sequence).
- `test_has_initial_state_decode_plus_new_prefill` — decode kernel + recurrent kernel interaction; decode-slot state survives the recurrent kernel zeroing prefill slots.

Eval - MMLU-Pro looks good

|       Tasks       |Version|    Filter    |n-shot|  Metric   |   |Value |   |Stderr|
|-------------------|------:|--------------|-----:|-----------|---|-----:|---|-----:|
|mmlu_pro           |      2|custom-extract|      |exact_match|↑  |0.8321|±  |0.0224|
| - biology         |      3|custom-extract|     5|exact_match|↑  |0.8500|±  |0.0819|
| - business        |      3|custom-extract|     5|exact_match|↑  |0.7500|±  |0.0993|
| - chemistry       |      3|custom-extract|     5|exact_match|↑  |0.8500|±  |0.0819|
| - computer_science|      3|custom-extract|     5|exact_match|↑  |0.8500|±  |0.0819|
| - economics       |      3|custom-extract|     5|exact_match|↑  |0.9000|±  |0.0688|
| - engineering     |      3|custom-extract|     5|exact_match|↑  |0.7500|±  |0.0993|
| - health          |      3|custom-extract|     5|exact_match|↑  |0.8000|±  |0.0918|
| - history         |      3|custom-extract|     5|exact_match|↑  |0.7500|±  |0.0993|
| - law             |      3|custom-extract|     5|exact_match|↑  |0.8000|±  |0.0918|
| - math            |      3|custom-extract|     5|exact_match|↑  |0.9500|±  |0.0500|
| - other           |      3|custom-extract|     5|exact_match|↑  |0.7000|±  |0.1051|
| - philosophy      |      3|custom-extract|     5|exact_match|↑  |0.9000|±  |0.0688|
| - physics         |      3|custom-extract|     5|exact_match|↑  |1.0000|±  |0.0000|
| - psychology      |      3|custom-extract|     5|exact_match|↑  |0.8000|±  |0.0918|

| Groups |Version|    Filter    |n-shot|  Metric   |   |Value |   |Stderr|
|--------|------:|--------------|------|-----------|---|-----:|---|-----:|
|mmlu_pro|      2|custom-extract|      |exact_match|↑  |0.8321|±  |0.0224|

### CODE REVIEW

File: tpu_inference/layers/common/gdn_attention.py [hunk 1]
```diff
@@ -168,9 +170,11 @@ def run_jax_gdn_attention_local(
         )
 
     if config.ragged_gated_delta_rule_impl == RaggedGatedDeltaRuleImpl.FUSED:
-        # The fused kernel does not yet honor `has_initial_state`. Fall
-        # through with the existing call signature; the chunked/ref paths
-        # used in production already mask via the new argument below.
+        # All three impls (fused / chunked / ref) honor `has_initial_state`
+        # — the fused recurrent kernel zeros h0 in VMEM for slots whose
+        # flag is 0. The branch is preserved (rather than unified into a
```
Comment (jrplatin) [inline tpu_inference/layers/common/gdn_attention.py]: Not sure I really understand this / if this is needed

Comment (qizzzh) [reply tpu_inference/layers/common/gdn_attention.py]: Ah stale comment during the development. Updated.

File: tests/kernels/fused_gdn_kernel_test.py [hunk 2]
```diff
@@ -220,6 +241,139 @@ def test_padded_max_num_req(self, decode_N, mixed_seqlens, extra_pad):
...
+                            check_dtypes=False)
+        # Compare the active slots; padding slots are not exercised.
+        self.assertAllClose(state_stale[:N],
+                            state_fresh[:N],
+                            atol=1e-4,
+                            rtol=1e-4,
+                            check_dtypes=False)
+
+    @parameterized.named_parameters(
+        # `all_true` covers the True-preserves-loaded-state path;
```
Comment (jrplatin) [inline tests/kernels/fused_gdn_kernel_test.py]: Don't really follow this either -- what does "True-preserves-loaded-state path" mean? Can you please simplify / make the wording here a bit more clear? Thanks!

File: tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py [hunk 3]
```diff
@@ -138,6 +145,14 @@ def fused_gdn(
         distribution: ``i32[2]`` — ``(decode_end, total)``.
         b: Raw betas ``[T, H_v]`` (sigmoid applied inside kernel).
             ``None`` means beta=1 (no beta gating).
+        has_initial_state: Boolean / int32 ``[max_num_req]``. ``True`` /
```
Comment (helloworld1) [inline tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py]: The has_initial_state has always been passing in as Boolean (i8) and also in chunked implementation, but I think kernel needs to convert to int32 for SMEM. Can we just document it as always boolean type so all implementations uses the same types?

**Review by helloworld1 — APPROVED** [review]

File: tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py [hunk 4]
```diff
@@ -187,6 +202,15 @@ def fused_gdn(
...
             jnp.float32)  # [H_v, num_lanes]
 
+    # Default to all-True (no masking), matching the pre-fix behaviour.
+    # Cast to int32 for SMEM compatibility — the recurrent kernel checks
+    # `has_init == 0` to decide whether to zero h0.
+    max_num_req = state_indices.shape[0]
+    if has_initial_state is None:
+        has_initial_state = jnp.ones((max_num_req, ), dtype=jnp.int32)
+    else:
+        has_initial_state = has_initial_state.astype(jnp.int32)
```
Comment (jrplatin) [inline tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py:213]: Why do we need to cast it? Aren't we initing it as INT32?

File: tpu_inference/kernels/gdn/fused_gdn_recurrent_kernel.py
Comment (jrplatin) [inline tpu_inference/kernels/gdn/fused_gdn_recurrent_kernel.py:1]: @yaochengji Can you PTAL at these changes? Thanks!

File: tests/kernels/fused_gdn_kernel_test.py (hunk 2 above)
Comment (qizzzh) [reply tests/kernels/fused_gdn_kernel_test.py]: Updated

File: tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py (hunk 3 above)
Comment (qizzzh) [reply tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py]: Updated

File: tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py (hunk 4 above)
Comment (qizzzh) [reply tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py:213]: As mentioned by @helloworld1 above, the kernel needs to convert it to int32 for SMEM

**Review by jrplatin — APPROVED** [review]

File: tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py (hunk 4 above)
Comment (helloworld1) [reply tpu_inference/kernels/gdn/fused_gdn_kernel_wrapper.py:213]: We need to cast because kernel's SMEM only support int32 or float32.

**Review by yaochengji — APPROVED** [review]
LGTM on the kernel change, thanks!
